pkg/controllers/state/cost.go

```diff
@@ -0,0 +1,370 @@
+/*
+Copyright The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package state
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	opmetrics "github.com/awslabs/operatorpkg/metrics"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/samber/lo"
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	crmetrics "sigs.k8s.io/controller-runtime/pkg/metrics"
+
+	v1 "sigs.k8s.io/karpenter/pkg/apis/v1"
+	"sigs.k8s.io/karpenter/pkg/cloudprovider"
+	"sigs.k8s.io/karpenter/pkg/metrics"
+)
+
+// NecessaryLabels defines the set of required Kubernetes labels that must be present
+// on NodeClaim objects for cost tracking to function properly.
+var NecessaryLabels = []string{corev1.LabelInstanceTypeStable, v1.CapacityTypeLabelKey, corev1.LabelTopologyZone, v1.NodePoolLabelKey}
+
+var (
+	CostTrackingErrorsTotal = opmetrics.NewPrometheusCounter(
+		crmetrics.Registry,
+		prometheus.CounterOpts{
+			Namespace: metrics.Namespace,
+			Subsystem: metrics.NodePoolSubsystem,
+			Name:      "cost_tracker_errors_total",
+			Help:      "Number of errors encountered during cost tracking operations. Labeled by nodepool and nodeclaim.",
+		},
+		[]string{
+			metrics.NodePoolLabel,
+		},
+	)
+)
+
+// ClusterCost tracks the cost of compute resources across all NodePools in a cluster.
+// This is an alpha-level component and its API may change without notice.
+//
+// The ClusterCost maintains real-time cost information by:
+// - Tracking NodeClaim additions and removals
+// - Managing instance type offerings and their prices
+// - Calculating aggregate costs per NodePool and cluster-wide
+//
+// All operations are thread-safe through internal locking mechanisms.
+type ClusterCost struct {
+	sync.RWMutex
+	npCostMap map[string]*NodePoolCost // nodepool.Name -> NodePoolCost
```

Comment:
nit: we shouldn't use resource names as a key in this map since it can introduce failures when resources are recreated. I'd use the NodePool UUID as the main key and if we need the name as well it can be stored alongside it.